# This script compares haplotypes generated by different phasing methods:
# 1. default during Sanger imputation on HRC (Eagle2)
# 2. SHAPEITv2 with 1000G and pedigree info
# 3. Ge's local phasing algorithm


### INIT
options(stringsAsFactors = F)
library(dplyr)
library(tidyr)
library(ggplot2)


### LOAD DATA
# From each dataset, extract one trio and one chromosome
# (trio selected so that all are on batch 24 and QC)
trio = c(fet="9981196059_R12C01", pat="9998714025_R08C02", mat="9998714025_R07C02")
bcfoptions = paste("-s", paste(trio, collapse=","), "-f '%POS %REF %ALT[ %GT]\n'")

in_eagle = "/mnt/HUNT/tsdphased/imputed_m24/22.vcf.gz"
in_shapeit = "/mnt/HUNT/tsdphased/m24chr22.phased.vcf.gz"
out_eagle = "/mnt/HUNT/haplotypes/test_eagle.vcf"
out_shapeit = "/mnt/HUNT/haplotypes/test_shapeit.vcf"

## extract the trio
# (sample order is same as in -s flag)
command = paste("bcftools query", in_eagle, bcfoptions, "-o", out_eagle)
print(command)
system(command)
command = paste("bcftools query", in_shapeit, bcfoptions, "-o", out_shapeit)
print(command)
system(command)

## load and split into haplotypes
eagle = read.table(out_eagle)
shapeit = read.table(out_shapeit)
colnames(eagle) = colnames(shapeit) = c("POS", "REF", "ALT", names(trio))
eagle = separate(eagle, fet, c("fetA", "fetB"), sep="\\|", convert=T) %>% 
	separate(pat, c("patA", "patB"), sep="\\|", convert=T) %>%
	separate(mat, c("matA", "matB"), sep="\\|", convert=T)
shapeit = separate(shapeit, fet, c("fetA", "fetB"), sep="\\|", convert=T) %>% 
	separate(pat, c("patA", "patB"), sep="\\|", convert=T) %>%
	separate(mat, c("matA", "matB"), sep="\\|", convert=T)

## flip (EAGLE dataset) alleles if needed
toflip = inner_join(eagle, shapeit, by=c("POS", "REF"="ALT", "ALT"="REF"))
toflip = paste(toflip$POS, toflip$REF, toflip$ALT, sep="_")
toflip = which(paste(eagle$POS, eagle$REF, eagle$ALT, sep="_") %in% toflip)
eagle[toflip, c("REF", "ALT")] = eagle[toflip, c("ALT", "REF")]
eagle[toflip, 4:9] = 1 - eagle[toflip, 4:9]

## identify transmission and merge
# (pat/mat identify the source of A hap)
phaseTable = expand.grid( fetA=0:1, fetB=0:1, pat=0:2, mat=0:2)
phaseTable$conclusion = "ME"
phaseTable$conclusion[c(1, 5, 13, 17:20, 24, 32, 36)] = "NI"
phaseTable$conclusion[c(7, 11, 14, 23, 26, 30)] = "mat"
phaseTable$conclusion[c(6, 10, 15, 22, 27, 31)] = "pat"

eagle = mutate(eagle, pat=patA+patB, mat=matA+matB) %>%
	inner_join(phaseTable, by=c("pat", "mat", "fetA", "fetB"))
shapeit = mutate(shapeit, pat=patA+patB, mat=matA+matB) %>%
	inner_join(phaseTable, by=c("pat", "mat", "fetA", "fetB"))
merged = full_join(eagle, shapeit, by=c("POS", "REF", "ALT"), suffix=c("e", "s"))


### PLOT
merged$conclusione = factor(merged$conclusione, c("ME", "mat", "pat", "NI"))
merged$conclusions = factor(merged$conclusions, c("ME", "mat", "pat", "NI"))
# as expected, most SNPs are non-informative (e.g. rare 00s):
filter(merged, !is.na(conclusions)) %>%
	ggplot(aes(x=POS)) +
	geom_point(aes(y=1, col=conclusions), pch="|") +
	geom_point(aes(y=2, col=conclusione), pch="|") +
	theme_bw()

# most of the rest seem to have paternal hap in A:
filter(merged, !is.na(conclusions), conclusions!="NI") %>%
	ggplot(aes(x=POS)) +
	geom_point(aes(y=1, col=conclusions), pch="|") +
	geom_point(aes(y=2, col=conclusione), pch="|") +
	theme_bw()
table(merged$conclusions)
table(merged$conclusione)
filter(merged, conclusione=="pat") %>% head
